carball/analysis/events/carry_detection.py

```diff
@@ -0,0 +1,187 @@
+                if hit_list[i].frame_number < start_frames[frame_index]:
+                    continue
+                if hit_list[i].frame_number > start_frames[frame_index]:
+                    # now we go backwards
+                    hit_number = i
+                    if hit_list[i].frame_number > end_frames[frame_index]:
+                        hit_number -= 1
+                    while (hit_number > 0 and hit_list[hit_number].player_id.id == player.id.id and
+                           hit_list[hit_number - 1].frame_number in valid_frames.index):
+                        hit_number -= 1
+                    # now we have the first hit that is in valid
+                    if hit_list[hit_number].frame_number < start_frames[frame_index]:
+                        # now we modify
+                        print('NEED MODIFICATION')
+
+            if starting_hit_index >= len(hit_list) - 1:
+                logger.warning("We have reached a point where dribbles should not exist anymore")
+        # Look for the first carry frame and find all hits before it by the same player
+        pass
+        """
+
+        return player_carry_data
+
+    def get_distance(self, player_frames: pd.DataFrame, start: int, end: int) -> pd.DataFrame:
+        x1 = player_frames.pos_x[start]
+        y1 = player_frames.pos_y[start]
+
+        x2 = player_frames.pos_x[end]
+        y2 = player_frames.pos_y[end]
+
+        return ((x2 - x1) ** 2 + (y2 - y1) ** 2) ** 0.5
+
+    def creat_start_end_frames(self, carry_frames: pd.DataFrame) -> Tuple[List[int], List[int]]:
+        carry_frame_index = carry_frames.index
+        shifted = carry_frame_index.to_series().shift(fill_value=0)
+        neg_shifted = carry_frame_index.to_series().shift(periods=-1, fill_value=0)
+        start_frames = carry_frame_index[abs(carry_frame_index - shifted) > 3]
+        end_frames = carry_frame_index[abs(carry_frame_index - neg_shifted) > 3]
+        return start_frames, end_frames
+
+    def add_carry_events(self, carry_data: CarryData,
+                         xy_distance: pd.DataFrame, player_frames: pd.DataFrame, player: Player, proto_game):
+
+        start_frames = carry_data.start_frames
+        end_frames = carry_data.end_frames
+        for i in range(len(carry_data.start_frames)):
+            total_time = carry_data.carry_frames.index.slice_locs(start_frames[i], end_frames[i])
+            carry_frames = carry_data.carry_frames.iloc[total_time[0]:total_time[1]]
+            total_time = carry_frames.game.delta.sum()
+            if total_time < 1:
+                continue
+
+            ball_carry = proto_game.game_stats.ball_carries.add()
+            ball_carry.start_frame_number = start_frames[i]
+            ball_carry.end_frame_number = end_frames[i]
+            ball_carry.player_id.id = player.id.id
+            ball_carry.carry_time = total_time
+            ball_carry.distance_traveled = self.get_distance(player_frames, start_frames[i], end_frames[i])
```

Comment:
Why don't you just sum up the ball position diffs?
`(ball.x.diff()**2 + ball.y.diff()**2).sqrt()` or something?

Reply:
This is straight line distance not along the path difference.
Though I might add that with player too!
Good suggestion